File: python/examples/readme_snippets.py
import numpy as np
import spectrograms as sg

# 1 second of 440 Hz sine wave
sample_rate = 16000
t = np.linspace(0, 1, sample_rate, dtype=np.float64)
samples = np.sin(2 * np.pi * 440 * t)


# Configure parameters
stft = sg.StftParams(n_fft=512, hop_size=256, window=sg.WindowType.hanning, centre=True)

params = sg.SpectrogramParams(stft, sample_rate=sample_rate)

# Compute power spectrogram
spec = sg.compute_linear_power_spectrogram(samples, params)

print(f"Shape: {spec.n_bins} bins x {spec.n_frames} frames")

stft = sg.StftParams(512, 256, sg.WindowType.hanning, True)
params = sg.SpectrogramParams(stft, 16000)

# Mel filterbank
mel = sg.MelParams(n_mels=80, f_min=0.0, f_max=8000.0)

# dB scaling
db = sg.LogParams(floor_db=-80.0)

# Compute mel spectrogram in dB
spec = sg.compute_mel_db_spectrogram(samples, params, mel, db)

# Access data
print(f"Mel bands: {spec.n_bins}")
print(f"Frames: {spec.n_frames}")
print(f"Frequency range: {spec.frequency_range()}")

signals = [
    np.random.randn(16000),
    np.random.randn(16000),
    np.random.randn(16000),
]

stft = sg.StftParams(512, 256, sg.WindowType.hanning, True)
params = sg.SpectrogramParams(stft, 16000)
mel = sg.MelParams(80, 0.0, 8000.0)
db = sg.LogParams(-80.0)

# Create plan once
planner = sg.SpectrogramPlanner()
plan = planner.mel_db_plan(params, mel, db)

# Reuse for all signals (much faster!)
for signal in signals:
    spec = plan.compute(signal)
    # Process spec...


# Create a 256x256 image
image = np.zeros((256, 256), dtype=np.float64)
for i in range(256):
    for j in range(256):
        image[i, j] = np.sqrt((i - 128) ** 2 + (j - 128) ** 2)

# Compute 2D FFT
spectrum = sg.fft2d(image)
print(f"Spectrum shape: {spectrum.shape}")
# Output: (256, 129) due to Hermitian symmetry

# Apply Gaussian blur via FFT
kernel = sg.gaussian_kernel_2d(9, 2.0)
blurred = sg.convolve_fft(image, kernel)

# Apply high-pass filter for edge detection
edges = sg.highpass_filter(image, 0.1)

# Compute power spectrum
power = sg.power_spectrum_2d(image)

images = [
    np.random.randn(256, 256),
    np.random.randn(256, 256),
    np.random.randn(256, 256),
]

# Create planner once
planner = sg.Fft2dPlanner()

# Reuse for all images (faster!)
for image in images:
    spectrum = planner.fft2d(image)
    power = np.abs(spectrum) ** 2
    # Process power spectrum...


stft = sg.StftParams(512, 160, sg.WindowType.hanning, True)
mfcc_params = sg.MfccParams(n_mfcc=13)

mfccs = sg.compute_mfcc(
    samples, stft, sample_rate=16000, n_mels=40, mfcc_params=mfcc_params
)

# Shape: (13, n_frames)
print(f"MFCCs: {mfccs.shape}")


stft = sg.StftParams(4096, 512, sg.WindowType.hanning, True)
chroma_params = sg.ChromaParams.music_standard()

chroma = sg.compute_chromagram(
    samples, stft, sample_rate=22050, chroma_params=chroma_params
)

# Shape: (12, n_frames)
print(f"Chroma: {chroma.shape}")

# Use class methods
window = sg.WindowType.hanning
kaiser = sg.WindowType.kaiser(beta=8.0)
gauss = sg.WindowType.gaussian(std=0.4)

# Speech processing preset
params = sg.SpectrogramParams.speech_default(sample_rate=16000)

# Music processing preset
params = sg.SpectrogramParams.music_default(sample_rate=44100)

spec = sg.compute_linear_power_spectrogram(samples, params)

# Dimensions
n_bins = spec.n_bins
n_frames = spec.n_frames

# Data (numpy array)
data = spec.data  # shape: (n_bins, n_frames)

# Axes
freqs = spec.frequencies
times = spec.times
f_min, f_max = spec.frequency_range()
duration = spec.duration()

# Original parameters
params = spec.params

File: examples/readme_snippets.rs
use non_empty_slice::NonEmptyVec;
use spectrograms::*;
use std::f64::consts::PI;

fn samples() -> NonEmptyVec<f64> {
    let sample_rate = 16000.0;
    let samples: Vec<f64> = (0..16000)
        .map(|i| {
            let t = i as f64 / sample_rate;
            (2.0 * PI * 440.0 * t).sin()
        })
        .collect();
    let samples = NonEmptyVec::new(samples).unwrap();
    samples
}

pub fn main() -> Result<(), SpectrogramError> {
    println!("---");
    println!("Basic spectrogram example:");
    println!("---");
    basic_spectrogram_example()?;

    println!("---");
    println!("Binaural spectrogram examples:");
    println!("---");
    binaural_examples()?;

    println!("---");
    println!("Mel spectrogram example:");
    println!("---");

    mel_spectrogram_example()?;

    println!("---");
    println!("Efficient batch processing example:");
    println!("---");

    efficient_batch_processing()?;

    println!("---");
    println!("2D FFT and image processing example:");
    println!("---");

    _2d_fft_and_image_processing()?;

    println!("---");
    println!("Efficient batch image processing example:");
    println!("---");

    efficient_batch_image_processing()?;

    println!("---");
    println!("MFCC example:");
    println!("---");

    mfccs()?;

    println!("---");
    println!("Chromagram example:");
    println!("---");

    _chromagram()?;

    println!("---");
    println!("MDCT example:");
    println!("---");

    mdct_example()?;

    Ok(())
}

fn basic_spectrogram_example() -> SpectrogramResult<()> {
    let samples = samples();
    // Configure parameters
    let stft = StftParams::new(
        nzu!(512),           // FFT size
        nzu!(256),           // hop size
        WindowType::Hanning, // window
        true,                // centre frames
    )?;

    let params = SpectrogramParams::new(
        stft, 16000.0, // sample rate
    )?;

    // Compute power spectrogram
    let spec = LinearPowerSpectrogram::compute(samples.as_non_empty_slice(), &params, None)?;

    println!("Shape: {} bins x {} frames", spec.n_bins(), spec.n_frames());
    Ok(())
}

fn mel_spectrogram_example() -> SpectrogramResult<()> {
    let samples = samples();
    let stft = StftParams::new(nzu!(512), nzu!(256), WindowType::Hanning, true)?;
    let params = SpectrogramParams::new(stft, 16000.0)?;

    // Mel filterbank
    let mel = MelParams::new(
        nzu!(80), // n_mels
        0.0,      // f_min
        8000.0,   // f_max
    )?;

    // dB scaling
    let db = LogParams::new(-80.0)?;

    // Compute mel spectrogram in dB
    let spec = MelDbSpectrogram::compute(&samples, &params, &mel, Some(&db))?;

    // Access data
    println!("Mel bands: {}", spec.n_bins());
    println!("Frames: {}", spec.n_frames());
    println!("Frequency range: {:?}", spec.axes().frequency_range());

    Ok(())
}

fn efficient_batch_processing() -> SpectrogramResult<()> {
    use non_empty_slice::non_empty_vec;
    use spectrograms::*;
    let signals = vec![
        non_empty_vec![0.0; nzu!(16000)],
        non_empty_vec![0.0; nzu!(16000)],
        non_empty_vec![0.0; nzu!(16000)],
    ];

    let stft = StftParams::new(nzu!(512), nzu!(256), WindowType::Hanning, true)?;
    let params = SpectrogramParams::new(stft, 16000.0)?;
    let mel = MelParams::new(nzu!(80), 0.0, 8000.0)?;
    let db = LogParams::new(-80.0)?;

    // Create plan once
    let planner = SpectrogramPlanner::new();
    let mut plan = planner.mel_plan::<Decibels>(&params, &mel, Some(&db))?;

    // Reuse for all signals (much faster!)
    for signal in signals {
        let _spec = plan.compute(&signal)?;
        // Process spec...
    }

    Ok(())
}

fn _2d_fft_and_image_processing() -> SpectrogramResult<()> {
    use ndarray::Array2;
    use spectrograms::fft2d::*;
    use spectrograms::image_ops::*;

    // Create a 256x256 image
    let image = Array2::<f64>::from_shape_fn((256, 256), |(i, j)| {
        ((i as f64 - 128.0).powi(2) + (j as f64 - 128.0).powi(2)).sqrt()
    });

    // Compute 2D FFT
    let spectrum = fft2d(&image.view())?;
    println!("Spectrum shape: {:?}", spectrum.shape());
    // Output: [256, 129] due to Hermitian symmetry

    // Apply Gaussian blur via FFT
    let kernel = gaussian_kernel_2d(spectrograms::nzu!(9), 2.0)?;

    let _blurred = convolve_fft(&image.view(), &kernel.view())?;

    // Apply high-pass filter for edge detection
    let _edges = highpass_filter(&image.view(), 0.1)?;

    // Compute power spectrum
    let _power = power_spectrum_2d(&image.view())?;

    Ok(())
}

fn efficient_batch_image_processing() -> SpectrogramResult<()> {
    use ndarray::Array2;
    use spectrograms::fft2d::Fft2dPlanner;

    let images = vec![
        Array2::<f64>::zeros((256, 256)),
        Array2::<f64>::zeros((256, 256)),
        Array2::<f64>::zeros((256, 256)),
    ];

    // Create planner once
    let mut planner = Fft2dPlanner::new();

    // Reuse for all images (faster!)
    for image in &images {
        let spectrum = planner.fft2d(&image.view())?;
        let _power = spectrum.mapv(|c| c.norm_sqr());
        // Process power spectrum...
    }

    Ok(())
}

fn mfccs() -> SpectrogramResult<()> {
    let samples = samples();
    let stft = StftParams::new(nzu!(512), nzu!(160), WindowType::Hanning, true)?;
    let mfcc_params = MfccParams::new(nzu!(13));

    let mfccs = mfcc(
        &samples,
        &stft,
        16000.0,
        nzu!(40), // n_mels
        &mfcc_params,
    )?;

    // Shape: (13, n_frames)
    println!("MFCCs: {} x {}", mfccs.nrows(), mfccs.ncols());
    Ok(())
}

fn _chromagram() -> SpectrogramResult<()> {
    let samples = samples();
    let stft = StftParams::new(nzu!(4096), nzu!(512), WindowType::Hanning, true)?;
    let chroma_params = ChromaParams::music_standard();

    let chroma = chromagram(&samples, &stft, 22050.0, &chroma_params)?;

    // Shape: (12, n_frames) - one row per pitch class
    println!("Chroma: {} x {}", chroma.nrows(), chroma.ncols());

    Ok(())
}

fn binaural_examples() -> SpectrogramResult<()> {
    use spectrograms::binaural::*;

    let sample_rate = 44100.0;
    let duration_secs = 2.0;
    let n_samples = (sample_rate * duration_secs) as usize;

    // Simulate stereo audio: left channel slightly delayed (ITD)
    let left: Vec<f64> = (0..n_samples)
        .map(|i| (2.0 * PI * 440.0 * i as f64 / sample_rate).sin())
        .collect();
    let right: Vec<f64> = (0..n_samples)
        .map(|i| (2.0 * PI * 440.0 * (i as f64 + 2.0) / sample_rate).sin())
        .collect();

    let left = non_empty_slice::NonEmptyVec::new(left).unwrap();
    let right = non_empty_slice::NonEmptyVec::new(right).unwrap();
    let audio = [left.as_non_empty_slice(), right.as_non_empty_slice()];

    let stft = StftParams::new(nzu!(4096), nzu!(1024), WindowType::Hanning, true)?;
    let params = SpectrogramParams::new(stft, sample_rate)?;

    // ITD spectrogram
    let itd_params = ITDSpectrogramParams::new(params.clone(), 50.0, 620.0, None)?;
    let mut plan = StftPlan::new(&params)?;
    let itd = compute_itd_spectrogram(audio, &itd_params, &mut plan)?;
    println!(
        "ITD: {} bins x {} frames, freq range: {:.0}-{:.0} Hz",
        itd.n_bins(),
        itd.n_frames(),
        itd.frequency_range().0,
        itd.frequency_range().1
    );

    // IPD spectrogram (wrapped phase)
    let ipd_params = IPDSpectrogramParams::new(params.clone(), 50.0, 620.0, true)?;
    let ipd = compute_ipd_spectrogram(audio, &ipd_params, &mut plan)?;
    println!(
        "IPD: {} bins x {} frames (radians)",
        ipd.n_bins(),
        ipd.n_frames()
    );

    // ILD spectrogram
    let ild_params = ILDSpectrogramParams::new(params.clone(), 1700.0, 4600.0)?;
    let ild = compute_ild_spectrogram(audio, &ild_params, &mut plan)?;
    println!(
        "ILD: {} bins x {} frames (dB)",
        ild.n_bins(),
        ild.n_frames()
    );

    // ILR spectrogram
    let ilr_params = ILRSpectrogramParams::new(params.clone(), 1700.0, 4600.0)?;
    let ilr = compute_ilr_spectrogram(audio, &ilr_params, &mut plan)?;
    println!(
        "ILR: {} bins x {} frames (range [-1, 1])",
        ilr.n_bins(),
        ilr.n_frames()
    );

    // Histograms
    let itd_hist = itd.histogram(None, None, false, true);
    println!(
        "ITD histogram: {} bins x {} frames",
        itd_hist.nrows(),
        itd_hist.ncols()
    );

    let ild_hist = ild.histogram(None, None, None, false, true);
    println!(
        "ILD histogram: {} bins x {} frames",
        ild_hist.nrows(),
        ild_hist.ncols()
    );

    Ok(())
}

fn mdct_example() -> SpectrogramResult<()> {
    use spectrograms::{MdctParams, imdct, mdct};

    let sample_rate = 44100.0;
    let freq = 440.0;
    let n_samples = 8192usize;
    let signal: Vec<f64> = (0..n_samples)
        .map(|i| (2.0 * PI * freq * i as f64 / sample_rate).sin())
        .collect();
    let signal = non_empty_slice::NonEmptyVec::new(signal).unwrap();

    // Sine window gives perfect reconstruction with 50% hop
    let params = MdctParams::sine_window(nzu!(512))?;
    println!(
        "Window: {} samples, {} coefficients per frame",
        params.window_size.get(),
        params.n_coefficients()
    );

    // Forward MDCT
    let coefficients = mdct(signal.as_non_empty_slice(), &params)?;
    println!(
        "MDCT shape: {} bins x {} frames",
        coefficients.nrows(),
        coefficients.ncols()
    );

    // Inverse MDCT with perfect reconstruction
    let reconstructed = imdct(&coefficients, &params, Some(n_samples))?;
    let max_err = signal
        .as_slice()
        .iter()
        .zip(reconstructed.iter())
        .skip(512)
        .take(n_samples - 1024)
        .map(|(a, b)| (a - b).abs())
        .fold(0.0f64, f64::max);
    println!("Max reconstruction error (interior): {max_err:.2e}");

    Ok(())
}

#[allow(unused)]
fn window_functions() -> SpectrogramResult<()> {
    // Parse from string
    let window: WindowType = "hanning".parse()?;
    let kaiser: WindowType = "kaiser=8.0".parse()?;

    // Or use constructors
    let hann = WindowType::Hanning;
    let gauss = WindowType::Gaussian { std: 0.4 };

    // Generate windows
    let hann_window = make_window(WindowType::Hanning, nzu!(512));
    let kaiser_window = make_window(WindowType::Kaiser { beta: 8.0 }, nzu!(512));
    // etc.

    Ok(())
}

#[allow(unused)]
fn default_presets() -> SpectrogramResult<()> {
    // Speech processing preset
    // n_fft=512, hop_size=160
    let params = SpectrogramParams::speech_default(16000.0)?;

    // Music processing preset
    // n_fft=2048, hop_size=512
    let params = SpectrogramParams::music_default(44100.0)?;

    Ok(())
}

#[allow(unused)]
fn accessing_results() -> SpectrogramResult<()> {
    let samples = samples();
    let params = SpectrogramParams::speech_default(16000.0)?;
    let spec = LinearPowerSpectrogram::compute(&samples, &params, None)?;

    // Dimensions
    let n_bins = spec.n_bins();
    let n_frames = spec.n_frames();

    // Data (ndarray::Array2<f64>)
    let data = spec.data();

    // Axes
    let freqs = spec.axes().frequencies();
    let times = spec.axes().times();
    let (f_min, f_max) = spec.axes().frequency_range();
    let duration = spec.axes().duration();

    // Original parameters
    let params = spec.params();

    Ok(())
}
